platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/transform/TransformationChainData.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.component.resolution.failure.transform;
+
+import com.google.common.collect.ImmutableList;
+import org.gradle.api.internal.attributes.ImmutableAttributes;
+
+import java.util.stream.Collectors;
+
+/**
+ * Represents a variant which is produced as the result of applying an artifact transform chain
+ * to a root producer variant.
+ * <p>
+ * Immutable data class.  Meant to be easily serialized as part of build operation recording and tracing.
+ */
+public final class TransformationChainData {
+    private final SourceVariantData startingVariant;
+    private final ImmutableList<TransformData> steps;
+
+    public TransformationChainData(SourceVariantData startingVariant, ImmutableList<TransformData> steps) {
+        this.startingVariant = startingVariant;
+        this.steps = steps;
+    }
+
+    public SourceVariantData getInitialVariant() {
+        return startingVariant;
+    }
+
+    public String describeTransformations() {
+        return steps.stream()
+            .map(TransformData::getTransformName)
+            .collect(Collectors.joining(", "));
+    }
+
+    public ImmutableAttributes getFinalAttributes() {
+        return steps.get(steps.size() - 1).getToAttributes();
```

Comment:
IMO if we're doing this, we should have an `assert !steps.isEmpty()` in the constructor

Reply:
As stated in the other comment, I think its more useful to store proxies for the `TransformStep`s instead of proxies for `TransformedVariant`s. 

For this reason, I think we should just pass these final attributes into the constructor here

Reply:
This is removed.